### ISSUE TITLE
Error importing .m3 file from Heroes into Blender 2.91

### ISSUE THREAD

**Issue by DerekX** [Talv/m3addon#32]
Import C:\Users\Derek\Models\Heroes of the Storm\Overwatch Tracer\storm_hero_tra
cer_myth20.m3
ERROR: Unknown section at offset 32 with tag=MODL version=30 repetitions=1 secti
onLengthInBytes=880 guessedUnusedSectionBytes=12 guessedBytesPerEntry=868.0
ERROR: Unknown section at offset 390848 with tag=MADD version=3 repetitions=1 se
ctionLengthInBytes=160 guessedUnusedSectionBytes=0 guessedBytesPerEntry=160.0
Python: Traceback (most recent call last):
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\__init__.py", line 4393, in execute
    m3import.importM3BasedOnM3ImportOptions(scene)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3import.py", line 1873, in importM3BasedOnM3ImportOptions
    importer.importM3BasedOnM3ImportOptions(scene)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3import.py", line 425, in importM3BasedOnM3ImportOptions
    self.model = m3.loadModel(fileName)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 1184, in loadModel
    sections = loadSections(filename, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 1127, in loadSections
    section.determineContentField(checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 45, in determineContentField
    self.content = self.structureDescription.createInstances(buffer=self.rawByte
s, count=indexEntry.repetitions, checkExpectedValue=checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 160, in createInstances
    return buffer[:count-1].decode("ASCII")
UnicodeDecodeError: 'ascii' codec can't decode byte 0xbd in position 26: ordinal
 not in range(128)

location: <unknown location>:-1

**Comment by Solstice245** [comment]
Hi, have you tried using [yretenai's branch of the m3addon?](https://github.com/yretenai/io_mesh_m3addon)

We haven't included support for those particular structures in this branch of the m3addon.
@Talv Do you know of any particular reason not to? If it doesn't hurt, perhaps we should.

**Comment by DerekX** [comment]
I am now, but I get this error when trying to activate it:

Traceback (most recent call last):
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\modules\ad
don_utils.py", line 351, in enable
    mod = __import__(module_name)
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\addons\io_
mesh_m3addon-develop\__init__.py", line 66, in <module>
    from .common import mlog
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\addons\io_
mesh_m3addon-develop\common.py", line 30, in <module>
    mlog = create_logger()
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\addons\io_
mesh_m3addon-develop\common.py", line 23, in create_logger
    os.makedirs(os.path.dirname(fname), exist_ok=True)
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\python\lib\os.py",
 line 223, in makedirs
    mkdir(name, mode)
PermissionError: [WinError 5] Access is denied: 'C:\\Program Files\\Blender Foun
dation\\Blender 2.91\\2.91\\scripts\\addons\\io_mesh_m3addon-develop\\logs'

**Comment by Solstice245** [comment]
Based on the error log, I'd suggest moving the script files to the AppData directory for that Blender version instead of the Program Files Blender directory or running Blender in admin mode, seeing as it's complaining about permissions. If that doesn't help, I'd recommend submitting an issue at the branch's respective repo.

**Comment by DerekX** [comment]
Tadaah! It worked, thanks much! :)

**Comment by Talv** [comment]
> Hi, have you tried using [yretenai's branch of the m3addon?](https://github.com/yretenai/io_mesh_m3addon)
> 
> We haven't included support for those particular structures in this branch of the m3addon.
> @Talv Do you know of any particular reason not to? If it doesn't hurt, perhaps we should.

The reason these changes aren't merged is because in its current implementation it MIGHT break exporting capabilities, or just change behavior of some things due to to changes to function `findNameAndMarkAsUsedLike`. You can lookup the diff here: https://github.com/SC2Mapster/m3addon/pull/21/files#diff-2fe16068e367d29a7746514225b626b23603f7be39d2c1c98ea976ead3c9c3a9L350

Or it might be completely fine. I haven't tested it thoroughly, and I'm not sure what effect it might have on other SC2 models, which is the primary reason why it wasn't merged.

You can review that PR, it doesn't have many changes, and if you don't see anything potentially breaking other SC2 models, then it's good to merge.

**Comment by DerekX** [comment]
Actually, I AM having an exporting problem, ending in "KeyError: 12".

**Comment by Talv** [comment]
Should be fixed in 7a6285fdf96b23732fcde263e7f901b8655b937e .

@DerekX uninstall any other forks of m3addon you might have installed and download latest version from this repository.

Also keep in mind that the model you're importing: `storm_hero_tracer_myth20.m3` uses new material type that's exclusive to Storm engine. And cannot be properly imported to Blender by m3addon, nor exported. It's also not supported by SC2.

So if your intention is to do something with that model for SC2, then you'll have to redo materials.

**Comment by Talv** [comment]
Also keep in mind that the model you're importing: `storm_hero_tracer_myth20.m3` uses new material type that's exclusive to Storm engine. And cannot be properly imported to m3addon, nor exported. It's also not supported by SC2.

So if your intention is to do something with that model for SC2, then you'll have to redo materials.

**Comment by DerekX** [comment]
Latest version still has this error:

Traceback (most recent call last):
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\modules\ad
don_utils.py", line 351, in enable
    mod = __import__(module_name)
ModuleNotFoundError: No module named 'm3addon-master'
Traceback (most recent call last):
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\modules\ad
don_utils.py", line 351, in enable
    mod = __import__(module_name)
ModuleNotFoundError: No module named 'io_mesh_m3addon-develop'
Traceback (most recent call last):
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\modules\ad
don_utils.py", line 351, in enable
    mod = __import__(module_name)
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\addons\m3a
ddon-7a6285fdf96b23732fcde263e7f901b8655b937e\__init__.py", line 38, in <module>

    from .common import mlog
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\addons\m3a
ddon-7a6285fdf96b23732fcde263e7f901b8655b937e\common.py", line 30, in <module>
    mlog = create_logger()
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\addons\m3a
ddon-7a6285fdf96b23732fcde263e7f901b8655b937e\common.py", line 23, in create_log
ger
    os.makedirs(os.path.dirname(fname), exist_ok=True)
  File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\python\lib\os.py",
 line 223, in makedirs
    mkdir(name, mode)
PermissionError: [WinError 5] Access is denied: 'C:\\Program Files\\Blender Foun
dation\\Blender 2.91\\2.91\\scripts\\addons\\m3addon-7a6285fdf96b23732fcde263e7f
901b8655b937e\\logs'

Maybe this addon doesn't work well with Blender 2.91?

**Comment by Solstice245** [comment]
If you would read the log, it clearly indicates that the error is a Windows error, meaning the operating system prevented Blender from performing some action. I'll state more explicitly what I said before. I recommend moving the script files to the %APPDATA%/Blender Foundation/Blender/2.91/scripts/addons folder.

**Comment by DerekX** [comment]
It's already in that folder. I could actually activate it if I run Blender as Administrator. I'm just curious, why do I have to run Blender as Administrator in order to use that script?

**Comment by Solstice245** [comment]
The log states the file path of the script in the error.

`File "C:\Program Files\Blender Foundation\Blender 2.91\2.91\scripts\addons\m3a
ddon-7a6285fdf96b23732fcde263e7f901b8655b937e\common.py", line 30, in
mlog = create_logger()`

**Comment by DerekX** [comment]
So did I extract the script to the right directory or not? My problem might otherwise be solved, but I wanna try again with those HotS models. So standby...

**Comment by Solstice245** [comment]
As to answer your question about why, in Windows, changes made in the C:\Program Files directory requires special access permissions. mLog is attempting to write to a file within that directory while not having the permissions to do so. Launching  in administrator mode makes Windows grant the program permission to write within that directory.

**Comment by Talv** [comment]
Here's a thing I don't understand - how does the add-on ends up in program files directory? Because add-ons installed via Blender interface should always end up in user's appdata directory. Did you place it there manually? If so, that's against the instructions, and kinds expected to fail.

Overall if that's common installation mistake, then this should and will be addressed, but it's completely unrelated to the current issue.

**Comment by DerekX** [comment]
Talv you're probably right. I just tried with HotS Sally Whitemane, exported to .m3 with warnings. And the warnings ring true, as trying to import that .m3 file causes the SC2 editor to crash. I'll check out the "appdata" directory.

**Comment by DerekX** [comment]
And you guys say there is a way to merge versions of this addon? I have three versions installed at this time.

**Comment by Talv** [comment]
If there was a reason in the past to use different versions, that reason is now obsolete. All known improvements and fixes have been incorporated to `SC2Mapster/m3addon`, and this is the only version you should have installed. Otherwise we won't be able to provide any support.

Also, be sure to install latest version (from 5 days ago) - it "fixes" the mentioned crash issue.

**Comment by DerekX** [comment]
OK that problem appears to be fixed, but now I encounter this error with the .m3a file for HotS Garrosh Hellscream:

Python: Traceback (most recent call last):
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\ui\base.py", line 154, in execute
    m3import.importM3BasedOnM3ImportOptions(scene)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3import.py", line 1905, in importM3BasedOnM3ImportOptions
    importer.importM3BasedOnM3ImportOptions(scene)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3import.py", line 419, in importM3BasedOnM3ImportOptions
    self.model = m3.loadModel(fileName)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 1184, in loadModel
    sections = loadSections(filename, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 1127, in loadSections
    section.determineContentField(checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 45, in determineContentField
    self.content = self.structureDescription.createInstances(buffer=self.rawByte
s, count=indexEntry.repetitions, checkExpectedValue=checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 175, in createInstances
    list.append(self.createInstance(buffer=buffer, offset=instanceOffset, checkE
xpectedValue=checkExpectedValue));
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 155, in createInstance
    return M3Structure(self, buffer, offset, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 246, in __init__
    self.readFromBuffer(buffer, offset, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 264, in readFromBuffer
    field.readFromBuffer(self, buffer, fieldOffset, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 650, in readFromBuffer
    value = self.structFormat.unpack_from(buffer, offset)[0]
struct.error: unpack_from requires a buffer of at least 100 bytes

location: <unknown location>:-1

**Comment by DerekX** [comment]
Also encountering this error with the .m3a file for HotS Elite Tauren Chieftain:

Python: Traceback (most recent call last):
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\ui\base.py", line 154, in execute
    m3import.importM3BasedOnM3ImportOptions(scene)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3import.py", line 1905, in importM3BasedOnM3ImportOptions
    importer.importM3BasedOnM3ImportOptions(scene)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3import.py", line 419, in importM3BasedOnM3ImportOptions
    self.model = m3.loadModel(fileName)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 1184, in loadModel
    sections = loadSections(filename, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 1127, in loadSections
    section.determineContentField(checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 45, in determineContentField
    self.content = self.structureDescription.createInstances(buffer=self.rawByte
s, count=indexEntry.repetitions, checkExpectedValue=checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 175, in createInstances
    list.append(self.createInstance(buffer=buffer, offset=instanceOffset, checkE
xpectedValue=checkExpectedValue));
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 155, in createInstance
    return M3Structure(self, buffer, offset, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 246, in __init__
    self.readFromBuffer(buffer, offset, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 264, in readFromBuffer
    field.readFromBuffer(self, buffer, fieldOffset, checkExpectedValue)
  File "C:\Users\Derek\AppData\Roaming\Blender Foundation\Blender\2.91\scripts\a
ddons\m3addon-master\m3.py", line 652, in readFromBuffer
    raise Exception("Expected that %sV%s.%s has always the value %s, but it was
%s" % (owner.structureDescription.structureName, owner.structureDescription.stru
ctureVersion, self.name, self.expectedValue, value))
Exception: Expected that PATUV4.unknown71538f05 has always the value b'\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', but it was b'\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x80\xb5\x8c\xd5\x1aB'

location: <unknown location>:-1

**Comment by Talv** [comment]
> HotS Garrosh Hellscream

Fixed in 785e3449fa26ddfe8c5a0da3dd748fbd61999891

> HotS Elite Tauren Chieftain

Already fixed in current version - please always make sure you've the newest version before you submit a bug.

I'm closing this one, as there's already too many different topics stacked and it's hard to follow. In the future try to make separate issue for every separate problem. And possibly provide filenames of models, rather than their in-game names. Because it was non obvious to me that `HotS Elite Tauren Chieftain` is located at `Storm_Hero_ETC_RequiredAnims.m3a`.
